### PR TITLE
Add faster approximate gamma sampler

### PR DESCRIPTION
**Pull request by Gattocrucco**
Follow up to issue #38141.

This PR adds a `method='exact'/'approximate'` parameter to `jax.random.(log)gamma`. `'exact'` (default) selects the current implementation, `'approximate'` selects a faster approximate implementation (ported from the bartz package).

Note: the gradients w.r.t. the shape parameter `alpha` are different between the two paths, and not because of approximation error, they are actually different numbers. The `'exact'` path is a rejection sampler so it doesn't actually have a gradient, it's defined with a conventional rule in a custom jvp rule, while the `'approximate'` path is actually differentiable w.r.t. `alpha`.

P.S.: I used Claude Opus 4.8 & Fable 5 in this PR, they where originally marked as coauthors in the commit as Claude Code will do by default, but Google's CLA check was failing for them, so I removed them from the commit message.

### CODE REVIEW

**Comment by Gattocrucco** [comment]
Ping @jakevdp (in case you forgot; I don't particularly care about finishing this now).

**Comment by jakevdp** [comment]
Thanks for submitting this! Is there any way we can quantify the magnitude of the bias for the sake of the documentation? Are there regimes in which the biased samples should not be trusted?

File: jax/_src/random/core.py [hunk 1]
```diff
@@ -1650,6 +1651,11 @@ def gamma(key: ArrayLike,
       produces a result shape equal to ``a.shape``.
     dtype: optional, a float dtype for the returned values (default float64 if
       jax_enable_x64 is true, otherwise float32).
+    method: optional, the sampling algorithm to use, either ``'exact'`` (the
+      default) or ``'approximate'``. The ``'exact'`` method is a rejection
+      sampler. The ``'approximate'`` method is loop-free and faster but carries
+      a small bias. The gradient w.r.t. `a` differs between the two methods
```
Comment (jakevdp) [inline jax/_src/random/core.py:1657]: I'm confused on the note about this gradient. The gradient of `a` will in general depend on the particular draws from the sampler. I would expect that the two methods will return different samples, so the gradient with respect to `a` will vary just as it would if you put in different keys.

Or perhaps this is just saying that the gradient with respect to `a` carries the same bias as the underlying samples? Could we quantify that difference somehow?

File: jax/_src/random/core.py [hunk 2]
```diff
@@ -1746,6 +1770,148 @@ def _gamma(key, a, shape, dtype, log_space=False) -> Array:
...
+  """
+  # compute at least in float32; don't use dtypes.promote_types, fails in strict mode
+  compute_dtype = np.dtype('float32') if dtypes.finfo(dtype).bits < 32 else dtype
+  if shape is None:
+    shape = np.shape(a)
+  else:
+    _check_shape("gamma", shape, np.shape(a))
+  a = lax.convert_element_type(a, compute_dtype)
+  if np.shape(a) != shape:
+    a = jnp.broadcast_to(a, shape)
```
Comment (jakevdp) [inline jax/_src/random/core.py:1892]: We recently added utilities for these kinds of shape checks; see the `_check_broadcast_shapes` utility used in other samplers.

**Comment by Gattocrucco** [comment]
> Thanks for submitting this! Is there any way we can quantify the magnitude of the bias for the sake of the documentation? Are there regimes in which the biased samples should not be trusted?

In the internal docstring I report a 0.05% max error on quantiles, from the literature. Anecdotally, the max error I observed while stress-testing the implementation was 0.06%, so it checks out. I will make a script that tests this systematically so I can report the actual max in the public docstring.

Explanation: "the max error on quantiles" is that, say you take the 20%tile of the distribution, it's off by 0.05% tops relative to the true 20% gamma quantile across all possible values of the distribution parameter alpha, and this holds for all quantiles.

File: jax/_src/random/core.py (hunk 1 above)
Comment (Gattocrucco) [reply jax/_src/random/core.py:1657]: This is a substantial thing, not just an approximation error!

The gradient of a random sampler w.r.t. the distribution parameters is not defined in general. Basic example to show this:

Consider $U\sim U(0,1)$, and $X(a)=aU\sim U(0,a)$. In this example, $U$ is the jax rng and $X(a)$ is a uniform distribution sampler. The derivative $\partial X/\partial a = U$ is well defined for this specific sampler.

But the sampler is not unique. We could sample $U(0,a)$ with a rejection sampler by drawing $Y_i\sim U(0,b)$, $b>a$, then $X'(a) = {}$ the first $Y_i<a$. This samples from $U(0,a)$, but $\partial X'/\partial a = 0$ almost always, but for when $a$ crosses an $Y_i$ that was rejected or the accepted one.

So $X(a)$ and $X'(a)$ are both valid samplers of $U(0,a)$, but the first has a derivative, while the second doesn't.

$X(a)$ is the analogue of the 'approximate' gamma sampler I implemented (actual derivable arithmetic) while $X'(a)$ is analogous to the 'exact' default rejection sampler.

The derivative of the 'exact' sampler is defined here with polite fiction:

https://github.com/jax-ml/jax/blob/7d5524fa772d99cbfb373fe7978935d1be25a696/jax/_src/lax/special.py#L534-L552

I haven't read the code in detail, but at a glance it looks like jax is pretending that the sampler is $X(\alpha) = F^{-1}(U;\alpha)$, with $F(\cdot;\alpha)$ the Gamma distribution cdf. This is sort of a "standardized form" for the sampler, but this derivative does not actually correspond to how the outputs of the sampler change as the input parameter changes.

File: jax/_src/random/core.py (hunk 2 above)
Comment (Gattocrucco) [reply jax/_src/random/core.py:1892]: Ok, will do.

File: jax/_src/random/core.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/random/core.py:1657]: Interesting, thanks for laying that out. I wonder if @mattjj has thought about this kind of thing?

Comment (Gattocrucco) [reply jax/_src/random/core.py:1657]: Additional note: I wondered if maybe, in the case of smooth samplers (no rejections), the derivative could be unique. Answer: it isn't even for samplers that are smooth in the parameters, if the sampler uses multiple sources of randomness (draws multiple RVs for each output sample).

Example: use $Z_1,Z_2\sim N(0,1^2)$ as source of randomness. $X(a) = aZ_1\sim N(0,a^2)$, and $X'(a) = a(Z_1 \sin a + Z_2 \cos a) \sim N(0, a^2)$. Then $\partial X/\partial a = X/a$, while $\partial X'/\partial a = X'/a + a (Z_1 \cos a - Z_2\sin a)$, so the derivatives are different at the same output value.

Comment (jakevdp) [reply jax/_src/random/core.py:1657]: Yeah, a simpler example is you sample normally distributed values `x`, then create another normally-distributed sample `-x`. These have the same distribution, but opposite gradients!

Comment (Gattocrucco) [reply jax/_src/random/core.py:1657]: Well, initially I was about to lead with a sign-flip example similar to yours, but then I switched to the more complex one above because the sign flip actually has the same derivative when the output sample happens to be equal! The sign flips _at given random seed_, but if you feed two different seeds to the original and flipped generators such as to make them output the same $x$, then the derivatives are also equal.

Comment (Gattocrucco) [reply jax/_src/random/core.py:1657]: Trying to close this conversation: I'd keep the current behavior of this PR, which is that the 'approximate' method has different (not just approximate) derivatives from the 'exact' method. Reason: the properties in expectation are the same, but it's nice that the derivatives of the 'approximate' method are actual derivatives of the samples instead of a value that will produce consistent estimates of derivatives of expected values when involved in averages, I think this wins over making the derivatives match.

Pro of different derivatives:
- possibly faster derivative in the approximate case, not measured though
- the derivative of the approximate case is an actual derivative
- the derivative of the approximate case matches the approximation used

Pro of same derivatives:
- the derivative always have the same value across the two methods when samples coincide

**Review by jakevdp — APPROVED** [review]
Sounds good – let's get this in!

**Comment by jakevdp** [comment]
Could you please rebase your changes on an updated main branch?

**Comment by Gattocrucco** [comment]
Follow up on measuring the accuracy of the approximation: this script (written with Claude Fable 5) measures the quantile accuracy of the internal function `_loggamma_chisquare`. The result confirms the bound reported in the literature (max relative gamma quantile error 0.06% measured vs. 0.05% literature in a 0.998 interval) and also shows the worst case at more extreme quantiles.

```python
import jax

jax.config.update("jax_enable_x64", True)

import numpy as np
import scipy.stats
from jax import numpy as jnp
from jax._src.random import core

# shapes >= 4: `_gamma_approx` only evaluates the series at a_user + 4, and the
# error decreases with the shape, so the values next to 4 matter most
a_grid = [4.0, 4.1, 4.25, 4.5, 5.0, 6.0, 8.0, 16.0, 100.0, 10_000.0]
n = 10_000_000
keys = jax.random.split(jax.random.key(20260717), len(a_grid))

p_central = 0.001
z_central = scipy.stats.norm.isf(p_central)

W = 12
print("error = abs error on log(gamma) = rel error on gamma to first order")
print(
  f"{'':>8} "
  + " ".join(
    f"{t:>{2 * W + 1}}"
    for t in (f"all {n:_} samples", f"p in [{p_central}, {1 - p_central}]")
  )
)
print(f"{'shape':>8} " + " ".join(2 * [f"{'max err':>{W}} {'at p':>{W}}"]))
for key, a in zip(keys, a_grid):
  aa = jnp.full((n,), a)
  log_sample = core._loggamma_chisquare(key, aa)
  z = core._normal(key, aa.shape, aa.dtype)  # the same normal drawn inside
  p = scipy.stats.norm.cdf(z)
  err = np.abs(log_sample - np.log(scipy.stats.gamma.ppf(p, a)))
  central = np.abs(z) <= z_central
  cells = []
  for masked in (err, np.where(central, err, 0.0)):
    i = np.argmax(masked)
    cells.append(f"{err[i]:>#{W}.2g} {p[i]:>#{W}.2g}")
  print(f"{a:>8g} " + " ".join(cells))
```

```
error = abs error on log(gamma) = rel error on gamma to first order
            all 10_000_000 samples       p in [0.001, 0.999]
   shape      max err         at p      max err         at p
       4        0.059      1.3e-07      0.00064       0.0010
     4.1         0.13      1.0e-08      0.00062       0.0010
    4.25        0.042      1.4e-07      0.00061       0.0010
     4.5        0.027      2.2e-07      0.00058       0.0010
       5        0.080      2.9e-09      0.00054       0.0010
       6        0.013      4.5e-08      0.00046       0.0010
       8       0.0023      9.1e-08      0.00034       0.0010
      16      0.00021      1.1e-06      0.00014       0.0010
     100      2.5e-05          1.0      2.2e-05         0.65
   10000      1.4e-06      1.3e-07      7.1e-07          1.0
```

**Comment by jakevdp** [comment]
That's a really useful summary of the stats – thanks for posting!
